distribution/tools/server-cli/src/main/java/org/elasticsearch/server/cli/ServerProcess.java

```diff
@@ -0,0 +1,259 @@
+/*
+ * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
+ * or more contributor license agreements. Licensed under the Elastic License
+ * 2.0 and the Server Side Public License, v 1; you may not use this file except
+ * in compliance with, at your election, the Elastic License 2.0 or the Server
+ * Side Public License, v 1.
+ */
+
+package org.elasticsearch.server.cli;
+
+import org.elasticsearch.bootstrap.BootstrapInfo;
+import org.elasticsearch.bootstrap.ServerArgs;
+import org.elasticsearch.cli.ExitCodes;
+import org.elasticsearch.cli.ProcessInfo;
+import org.elasticsearch.cli.Terminal;
+import org.elasticsearch.cli.UserException;
+import org.elasticsearch.common.io.stream.OutputStreamStreamOutput;
+import org.elasticsearch.core.IOUtils;
+import org.elasticsearch.core.PathUtils;
+import org.elasticsearch.core.SuppressForbidden;
+
+import java.io.IOException;
+import java.io.OutputStream;
+import java.io.UncheckedIOException;
+import java.nio.file.Files;
+import java.nio.file.Path;
+import java.nio.file.Paths;
+import java.nio.file.attribute.FileAttribute;
+import java.util.ArrayList;
+import java.util.HashMap;
+import java.util.List;
+import java.util.Map;
+
+import static org.elasticsearch.server.cli.ProcessUtil.nonInterruptible;
+
+/**
+ * A helper to control a {@link Process} running the main Elasticsearch server.
+ *
+ * <p> The process can be started by calling {@link #start(Terminal, ProcessInfo, ServerArgs, Path)}.
+ * The process is controlled by internally sending arguments and control signals on stdin,
+ * and receiving control signals on stderr. The start method does not return until the
+ * server is ready to process requests and has exited the bootstrap thread.
+ *
+ * <p> The caller starting a {@link ServerProcess} can then do one of several things:
+ * <ul>
+ *     <li>Block on the server process exiting, by calling {@link #waitFor()}</li>
+ *     <li>Detach from the server process by calling {@link #detach()}</li>
+ *     <li>Tell the server process to shutdown and wait for it to exit by calling {@link #stop()}</li>
+ * </ul>
+ */
+public class ServerProcess {
+
+    // the actual java process of the server
+    private final Process jvmProcess;
+
+    // the thread pumping stderr watching for state change messages
+    private final ErrorPumpThread errorPump;
+
+    // a flag marking whether the java process has been detached from
```

Comment:
This comment reads like it's been truncated.

Reply:
Just poor grammar on my part. I've reworded.